### ISSUE TITLE
Serialize collection of polymorphic objects

### ISSUE THREAD

**Issue by cazacutudor** [jsonapi-serializer/jsonapi-serializer#121]
## Problem

I want to serialize a collection that's containing polymorphic objects, how can i do that ? It is supported ?

# Example

CollectionSerializer.rb
```
class CollectionSerializer
      include JSONAPI::Serializer
end
```

MovieSerializer.rb
```
class MovieSerializer
      include JSONAPI::Serializer

      attributes :name
end
```

ShowSerializer.rb
```
class ShowSerializer
      include JSONAPI::Serializer

      attributes :name
end
```

LiveShowSerializer.rb
```
class LiveShowSerializer
      include JSONAPI::Serializer

      attributes :name
end
```

controller.rb
```
data = [
    <Api::Entities::Movie id=10 name="Zoro" position=1>, 
    <Api::Entities::Show id=20 name="Zoro si Precini" position=2>,
    <Api::Entities::LiveShow id=30 name="Zoro si Dus-manii" position=3>
]

CollectionSerializer.new(data).serializable_hash.to_json

# expected output
"data": [
#   {
#     "id": "1",
#     "type": "movie",
#     "attributes": {
#       "name": "Zoro",
#     }
#   },
#   {
#       "id": "2",
#       "type": "show",
#       "attributes": {
#         "name": "Zoro si Precini",
#       }
#   },
#   {
#     "id": "3",
#     "type": "live_show",
#     "attributes": {
#       "name": "Zoro si Dus-manii",
#     }
#   }
# ]
```

## Specifications

  - Version: 2.1.0
  - Ruby version: 2.6.6

**Comment by stas** [comment]
Hi there @cazacutudor 

Would you give it a chance first and review our documentation on how to enable the polymorphic relationships?
https://github.com/jsonapi-serializer/jsonapi-serializer#customizable-options (at the end of the table)

Let us know what didn't work :bow:

**Comment by cazacutudor** [comment]
Hi @stas :) , thank you for your advice. I already checked the documentation, as we use polymorphic relationships with success in our production apps. Anyway, what i'm trying to figure out is: if is possible to serialize a collection of polymorphic object outside of an association (relationship), considering the polymorphic objects as root resources.

**Comment by stas** [comment]
@cazacutudor apologies for misreading your initial request :see_no_evil: 



> I want to serialize a collection that's containing polymorphic objects, how can i do that ? It is supported ?

In a nutshell, [nope](https://github.com/jsonapi-serializer/jsonapi-serializer/blob/master/docs/performance_methodology.md). And this is definitely something to reconsider, since the initial design/architecture didn't focus on that.

An alternative way you can approach this, is by manually handle the serialization of every record, something like:
```
data.map {|d1| resolve_serializer_class(d1).new(d1).serializable_hash }
```

Or still use a root serializer with a poly `has_many`, but return just [the `included` part](https://github.com/jsonapi-serializer/jsonapi-serializer#compound-document).

**Comment by ChrisHampel** [comment]
I believe this PR from the old repo was attempting to add this support https://github.com/Netflix/fast_jsonapi/pull/410, possibly could see if he would like to contribute it to this repo

**Comment by cazacutudor** [comment]
Thank you @stats, with some tweaks, i sure the provided alternatives would work :man_cartwheeling: 

Anyway, @ChrisHampel (thanks Chirs :pray:) have a very good point, if you don't mind, should we ask David if he wants to contribute to this repo.

**Comment by stas** [comment]
@ChrisHampel @cazacutudor of course, I'll be happy to review any work regarding this issue! My only concern is that it could lead to a bunch of performance issues, but let's keep our minds open! <3

**Comment by dvandersluis** [comment]
Hey guys thanks for pinging. My PR on `fast_jsonapi` never got merged obviously but we've been using it in production for a year and a half now, and haven't had issues with it (albeit I created it for our specific use case so there might be edge cases there).

I'm happy to rework it on top of this repo, but I need to evaluate first if we can drop in `jsonapi-serializer` instead of the older gem in our codebase before spending a ton of time on it.

**Comment by stas** [comment]
@dvandersluis I'd be curious to run some benchmarks to see how fast would be such a change. I'll take a look at the PR in the meantime.

Thank you!

**Comment by dvandersluis** [comment]
@stas @cazacutudor @ChrisHampel I have ported my PR to this gem.

**Comment by christophersansone** [comment]
@cazacutudor @stas This is indeed possible today, just FYI, by passing a `Proc` as the `serializer` parameter of the relationship.  The [relationship options documentation](https://github.com/jsonapi-serializer/jsonapi-serializer#customizable-options) states that a Proc is possible here, but there isn't a concrete example to demonstrate how to use it.

```ruby
has_many :widgets, serializer: -> { |o| "#{o.class.name}Serializer".constantize }
```

If you need to use a similar method often, you can create something more reusable, such as:

```ruby
class BaseSerializer
  include FastJsonapi::ObjectSerializer

  def self.serializer_by_class_name
    @serializer_by_class_name ||= Proc.new do |object|
      "#{object.class.name}Serializer".constantize
    end
  end
end

class ThingSerializer < BaseSerializer
  has_many :widgets, serializer: serializer_by_class_name
end
```

I have used this feature quite a bit, and it works like a charm.

**Comment by jrochkind** [comment]
I am interested in this feature!

Do I understand correctly that: PR is not feasible/welcome for this against 2.x; it is planned for 3.x; but there is no ETA and currently no work going on on 3.x?

if so, alternately, does anyone have a monkey-patch that succesfully adds this to 2.x? Some of the discussion above seemed to imply that perhaps someone did?

Or is there any chance of reconsidering #127 for 2.x?

**Comment by dvandersluis** [comment]
@jrochkind we are still using the fork I made for that PR (and have been for a few years) in production. I haven't touched it since more than a year ago and it's not fully up to date with master here but you could use it if you wanted: https://github.com/pexels/jsonapi-serializer/tree/heterogeneous-collection

**Comment by stas** [comment]
> Do I understand correctly that: PR is not feasible/welcome for this against 2.x; it is planned for 3.x; but there is no ETA and currently no work going on towards 3.x?

It's taken care in the #141 
https://github.com/jsonapi-serializer/jsonapi-serializer/blob/rewrite/spec/integration/mixed_collection_spec.rb

**Comment by jrochkind** [comment]
@stas Thanks! My understanding is that #141 is the 3.0 release, which is currently not having much progress due to lack of time/interest?  From your last comments on #141 five months ago, I conclude that we should not really expect a 3.0 release on any particular timeline, like not necessarily in 2022, or in even 2023? 

But I may be misunderstanding, that's why I ask! 

If there are no plans for #141 to be merged and released in the foreseeable future, I wonder if you'd reconsider a merge of #127 to a 2.x branch, since the comments on there seem to suggest it was deemed acceptable, but closed only because at that time about two years ago, you thought a 3.0 release was imminent?

**Comment by stas** [comment]
@jrochkind I've handled heterogeneous serialization in a much cleaner way without forking the v2. Eg.:

```ruby
class PolyRecordSerializer < BaseSerializer
  # Patched method to handle poly-ish record class
  #
  # TODO: Remove with v3 of the serializers
  #
  # @return [Hash]
  def self.record_hash(record, fieldset, includes_list, params = {})
    rhash = super
    rhash[:type] = record.class.name.underscore
    # Or dig even into relationships ...
    # rhash[:relationships][:to][:data][:type]
    rhash
  end
end
```

Which suggests I'm not interested in merging and maintaining the PR for two reasons:
 * the PR doesn't provide an API to handle a collection of actual heterogeneous records
 * the solution is hackish: `serializer_for()` has a ton of issues

---

> From your last comments on #141 five months ago, I conclude that we should not really expect a 3.0 release on any particular timeline, like not necessarily in 2022, or in even 2023?

Guys, nobody even bothered to test or run that brunch locally. I received 0 feedback regarding the API/interface of the v3. There's some experimental approach to tracking/resolving the serializers which I'm not 100% sure covers all the edge-cases. Documentation could use some love.

Now knowing all of this, I'm really trying to answer for myself, why should I push for the next release, even though it's 90+% ready and then have to face blame from folks who expect everything to be _just working_?

So yep, at the moment, I don't have a timeline when v3 is going to be ready. Seeing some support would definitely change my opinion  :hugs:

**Comment by jrochkind** [comment]
Thanks that example is super helpful! I will experiment with it. I don't totally understand how it works, but I'll play with it! Glad it is now here on this ticket for other interested parties to see too!  It does look far cleaner than other solutions, if it works It would be a nice thing to maybe document somewhere, since others have not come up with it on their own, not being as familiar with the possibilities as you! 

I'm sorry you haven't been getting the kind of support you'd like on this project, that sounds rough. Open source can be hard sometimes. 

I know you don't really want to say this, but I am leaning toward thinking consumers should consider this project basically in a maintenance-only mode, where new features are unlikely in either 2.x or 3.x. Fortunately, it seems to be pretty stable and working well in it's current state.

**Comment by stas** [comment]
Thanks @jrochkind 



> I know you don't really want to say this, but I am leaning toward thinking consumers should consider this project basically in a maintenance-only mode, where new features are unlikely in either 2.x or 3.x unless something changes. Fortunately, it seems to be pretty stable and working well in it's current state.

My real message to users is actually along these lines: we're running this software in production for years, and those are our own products or partners using it. If anything blows up, it's probably going to be fixed before others are even affected. The v2 is fully covering the serialization part of the JSON:API v1 spec, so yeah, that's in maintenance mode because there's no new stuff to add to it (`jsonapi.rb` is a great complement to that).

**Comment by jrochkind** [comment]
That makes sense. I'm not worried about anything blowing up. I just could really use the feature in this PR. 

I guess my options are....

* Use something other than `jsonapi-serializer` to create the collection serialization for a list of hetereogenous objects... I could still call out to `jsonapi-serializer` to serialize the individual objects. This might make sense, just create my own simple thing... it probably couldn't support `includes`, but maybe `fields`. 

* Use a fork of `jsonapi-serializer` that adds this feature, either create a new one, or use someone else's fork such as the one pointed out above

* Figure out a way to apply the feature as a "monkey patch" on top of current 2.x `jsonapi-serializer` instead of using a fork. @dvandersluis 's fork actually only changes a few lines of code, might be reasonable to apply it as a monkey-patch instead, which might be more convenient for keeping things in sync with upstream. 

* Use the unreleased `rewrite` branch of `jsonapi-serializer` directly from git, which reportedly has this feature.  There isn't a README in the diff, but maybe I could figure out how to access this new feature from the diff. 

Or, of course, just... don't try to serialize collections of polymorphic objects... or don't use this gem at all!

**Comment by stas** [comment]
@jrochkind any fork/monkey patch is partial tbh, I don't know of any proper fix, except for what's in v3. I already mentioned potential issues above:


> * the PR doesn't provide an API to handle a collection of actual heterogeneous records
> 
> * the solution is hackish: `serializer_for()` has a ton of issues



v3 has ~90+% coverage btw and proper fixtures. The tests are written more like integation tests, so it's all high-level:
https://github.com/jsonapi-serializer/jsonapi-serializer/blob/rewrite/spec/fixtures/movie.rb#L46-L99

> Or, of course, just... don't try to serialize collections of polymorphic objects... or don't use this gem at all!

Obviously :)

**Comment by jrochkind** [comment]
> the PR doesn't provide an API to handle a collection of actual heterogeneous records

It's not just `MySerializer.new(collection_of_actual_hetereogenous_records).serializable_hash`, where `MySerializer` implements `serializer_for`? 

"Is hackish has a ton of issues", I guess I'll take your word for it!  But like you say about this main gem, apparently some people have been using it for years and it's behaved solidly for their use cases. 

For someone that needs to serialize hetereogenous collection...  do you have any advice?

**Comment by christophersansone** [comment]
@jrochkind No need to monkey patch the entire gem or wait for a new release.  Each serializer is a class, where methods can be overridden as needed.  Just define a serializer class that does what you need it to do. @stas's example above is a good way to handle it if all you need to do is change the `type` being outputted.  My approach would be to simply build a `CollectionSerializer` that sent each object to its own serializer class for serialization, just like it does for relationships.  Something like this (untested):

```ruby
class CollectionSerializer
  include JSONAPI::Serializer

  def record_hash(record, *args)
    serializer_for(record).record_hash(record, *args)
  end

  def get_included_records(record, *args)
    serializer_for(record).get_included_records(record, *args)
  end
end
```

And as @stas mentions, `serializer_for` doesn't work in every situation, but that's okay!  Why?  Because it can be overridden as needed.  So if you implement this and then discover that it's not properly detecting the correct serializer, then just write a version of `serializer_for` that suits your needs.  This gem has been crafted to make these kinds of overrides simple and straightforward.  It's not monkey patching, it's simply a matter of implementing a custom approach in an isolated way when the default approach doesn't meet your requirements.

(@stas Thanks for the continued work on this!  I'm still happily on v1.x.  If it works, it works, so I also affirm not needing to perpetually push new versions and features when it does the job it sets out to do.)

**Comment by jrochkind** [comment]
@christophersansone I have been experimenting with stuff like that. But it's tricky to figure out what methods to override. 

`record_hash` and `get_included_records ` are both actually a _class_ method, so you can't override it in an instance method as in your example. 

I have experimented with over-riding `self.record_hash`, it kind of works, but no longer for instance seems to respect the `fields:` option. Which is maybe ok, I don't know.  I'm not sure if `include:` will really work even if you do override `get_included_records`, but maybe?

So yeah, figuring out exactly what to override to get it to mostly work is the whole thing... If someone has something that works well, it would be awesome not to have to re-invent the wheel. One thing that has been offered for that is the existing `serializer_for` fork of course.

**Comment by christophersansone** [comment]
@jrochkind Thanks, you're right about the class methods.  I updated the example to override the class methods.

From my read of the code, I don't see why `fields` would not work in my example.  The `fields` option is being memoized as `@fieldsets`, which is then passed into `record_hash`.  Can you confirm that it does not work and possibly try to find the root cause?

And yes, it looks like `include` won't work as is because it's validating that the relationship is defined on this specific serializer instead of the item's serializer.  In general, `includes` probably opens up is own can of worms, since it would seem the relationships wanting to be included will often be heterogeneous across the different record types.  I'm confident that this `CollectionSerializer` could be adapted to support it without much effort, but I imagine you'll want to define how you want to specify `include` and then adapt the serializer class to support it.

One of the main reasons why this gem has stood the test of time is that it is highly adaptable.  If there is anything that doesn't quite meet any individual needs, there is usually a configurable solution, whether through a lambda or Proc, or building an isolated override.  At the end of the day, each serializer is just a Plain Old Ruby Object, so any specific serializer can be wildly customized without impacting the others.

**Comment by jrochkind** [comment]
Yeah, I verified that `fields` did not work, but haven't not diagnosed why; thanks for finding more there. 

I could mess with trying to over-ride hash_for_collection too -- that one would require copy-and-paste-modify some code though.  And then I might find some other issue... I'm inclined to use the existing fork that is working for people rather than try to re-invent the wheel, as we see here it's not totally trivial, it seems odd that we'd each be re-inventing it in isolation instead of establishing a common solution. 

I'm curious if @stas has any advice for those who need this feature. But otherwise, sure, just investigating all my options, including other gems, just trying to figure out how to do what I need!

**Comment by stas** [comment]
> I'm curious if @stas has any advice for those who need this feature. But otherwise, sure, just investigating all my options, including other gems, just trying to figure out how to do what I need! It may just be that this gem will not do that.

My advise is to give #141 a shot. It does require a bit of hacking, but at least you'd invest in a future maintainable release rather than us continuing to support v2 with duct tape :)
